### PR TITLE
Ubuntu 24.04 & Nginx 1.24.0 compat work

### PR DESCRIPTION
**Pull request by cpu**
Ctz and I both riffed on this branch, working towards initial support for Nginx on Ubuntu 24.04. 

At a high-level this involves:

* Updating the testing infrastructure to add a Dockerfile and associated bits for testing on Ubuntu 24.04. This LTS release is brand spanking new and so there are no GitHub runner images available for it at this time.
* Implementing some `SSL_SESSION` manipulation APIs
* Implementing `SSL_get_SSL_CTX`
* Implementing `SSL_version`
* Stubbing out the `SSL_CONF_*` API surface (this will likely deserve revisiting soon to implement some config manipulation we could support)
* Stubbing out `SSL_sendfile` (this is kTLS specific in the OpenSSL API)

With these changes in place the Nginx 1.24.0 install is operable in a basic form with Rustls acting as libssl.

### CODE REVIEW

**Review by ctz — APPROVED** [review]
final 4 commits lgtm!

File: rustls-libssl/src/entry.rs [hunk 1]
```diff
@@ -1579,6 +1579,18 @@ entry! {
...
 }
 
+entry! {
+    pub fn _SSL_SESSION_set1_id_context(
+        sess: *mut SSL_SESSION,
+        sid_ctx: *const c_uchar,
+        sid_ctx_len: c_uint,
+    ) -> c_int {
+        let slice = try_slice!(sid_ctx, sid_ctx_len);
+        try_clone_arc!(sess).get_mut().set_context(slice);
```
Comment (cpu) [inline rustls-libssl/src/entry.rs:1608]: The upstream docs note the `sid_ctx_len` must be less than [`SSL_MAX_SID_CTX_LENGTH`](https://github.com/openssl/openssl/blob/067fbc01b9e867b31c71091d62f0f9012dc9e41a/include/openssl/ssl.h.in#L63)

Should we uphold that invariant in the event we implement `SSL_SESSION_get0_id_context` and don't want to break a downstream assumption?

**Review by cpu — COMMENTED** [review]
I can't "approve" since I opened the PR, but I've reviewed the early commits and do approve :ballot_box_with_check: Just one comment to ask about.

File: rustls-libssl/src/entry.rs (hunk 1 above)
Comment (cpu) [reply rustls-libssl/src/entry.rs]: Thanks!

**Review by cpu — COMMENTED** [review]
Latest [diff](https://github.com/rustls/rustls-openssl-compat/compare/e974b84a68e32e300e8e34b35364fb157a52950c..7e2202dda1d420d5f73a23dae29c9d2279faa46b) LGTM

**Comment by simondeziel** [comment]
>     * Updating the testing infrastructure to add a Dockerfile and associated bits for testing on Ubuntu 24.04. This LTS release is brand spanking new and so there are no GitHub runner images available for it at this time.

`ubuntu-24.04` is available now (https://github.com/actions/runner-images/issues/9848). `ubuntu-latest` still points to `ubuntu-22.04` for the time being though.

**Comment by cpu** [comment]
Thanks for the note, appreciated :-) We should look at adding that in to CI.

**Comment by cpu** [comment]
> I'll keep an eye on that upstream issue and try again once it has been resolved.

FWIW I rolled this into https://github.com/rustls/rustls-openssl-compat/pull/29 because I needed nginx 1.24+ for an integration test in that branch.

**Pull request merged**